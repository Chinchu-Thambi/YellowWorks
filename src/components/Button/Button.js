import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { Link } from 'gatsby';
import { pathOr } from 'ramda';

import theme from '../../util/theme';

// Button variants
const variants = {
  primary: {
    background: theme.palette.brand[0],
    text: theme.palette.contrast[0],
    hoverBackground: theme.palette.contrast[0],
    hoverText: theme.palette.base[0],
    outline: theme.palette.contrast[0],
  },
  secondary: {
    background: theme.palette.base[2],
    text: theme.palette.accent[5][0],
    hoverBackground: theme.palette.accent[5][0],
    hoverText: theme.palette.accent[5][4],
    outline: theme.palette.accent[5][0],
  },
  tertiary: {
    background: theme.palette.base[0],
    text: theme.palette.accent[5][0],
    hoverBackground: theme.palette.accent[5][0],
    hoverText: theme.palette.accent[5][4],
    outline: theme.palette.accent[5][0],
  },
  dark: {
    background: theme.palette.contrast[0],
    text: theme.palette.contrast[5],
    hoverBackground: theme.palette.contrast[1],
    hoverText: theme.palette.base[5],
  },
  link: {
    background: 'transparent',
    text: theme.palette.accent[3][0],
    hoverBackground: 'transparent',
    hoverText: theme.palette.accent[3][0],
    outline: 'transparent',
  },
  danger: {
    background: theme.palette.accent[0][0],
    text: theme.palette.base[0],
    hoverBackground: 'red',
    hoverText: theme.palette.base[0],
  },
};

// Button sizes
const sizes = {
  sm: `${theme.space[2]} ${theme.space[3]}`,
  md: `${theme.space[3]} ${theme.space[5]}`,
  lg: `${theme.space[3]} ${theme.space[7]}`,
};

const BaseButton = styled.button`
  font-family: ${({ variant }) => (variant === 'link' ? theme.fonts.sans : theme.fonts.title)};
  font-weight: ${({ variant }) => pathOr('400', [variant, 'fontWeight'], variants)};
  user-select: none;
  text-transform: uppercase;
  transition: background .12s ease-in-out;
  border: 1px solid transparent;
  text-decoration: none;
  line-height: initial;

  /* Control rounded borders */
  border-radius: 80px;

  /* Align all the child elements to the center */
  display: inline-flex;
  align-items: center;
  justify-content: center;
  text-align: center;

  /* Handle disabled */
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  opacity: ${(props) => props.disabled && 0.4};

  /* Handle size */
  padding: ${({ size }) => (size ? sizes[size] : sizes.md)};

  /* Handle size for  variant link */
  ${({ variant }) => variant && variant === 'link' && (`
    font-weight: ${theme.fontWeight[1]};

    :hover {
      text-decoration: underline;
    }
  `)}

  /* Handle font size */
  font-size: ${({ size }) => (size ? size === 'sm' && '14px' : '16px')};

  /* Handle outline */
  ${({ outline, variant }) => outline && (`
    border-color: ${variant ? variants[variant].outline : variants.primary.text};
  `)}

  /* Handle variants */
  ${({ variant }) => (`
    background: ${variant ? variants[variant].background : variants.primary.background};
    color: ${variant ? variants[variant].text : variants.primary.text};
    font-weight: ${variant && variants[variant]?.fontWeight};
  `)}

  /* Handle event, apply if button is not disabled */
  ${({ disabled, variant }) => !disabled && (`
    :hover,
    :active,
    :focus {
      background: ${variant ? variants[variant].hoverBackground : variants.primary.hoverBackground};
      color: ${variant ? variants[variant].hoverText : variants.primary.hoverText};
    }
  `)}
`;

const Button = React.forwardRef((props, ref) => {
  const {
    href, to, ...rest
  } = props;

  let element = 'button';
  if (href) element = 'a';
  if (to) element = Link;

  // eslint-disable-next-line react/jsx-props-no-spreading
  return <BaseButton ref={ref} as={element} href={href} to={to} {...rest} />;
});

export default Button;

Button.propTypes = {
  variant: PropTypes.oneOf([
    'primary',
    'secondary',
    'tertiary',
    'dark',
    'link',
    'danger',
  ]),
  outline: PropTypes.string,
  disabled: PropTypes.bool,
  type: PropTypes.string,
  size: PropTypes.oneOf([
    'sm', 'md', 'lg',
  ]),
  href: PropTypes.string,
  to: PropTypes.string,
};

Button.defaultProps = {
  variant: undefined,
  outline: undefined,
  disabled: false,
  type: undefined,
  size: undefined,
  href: undefined,
  to: undefined,
};
