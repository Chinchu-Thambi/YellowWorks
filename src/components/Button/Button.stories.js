/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { text, select } from '@storybook/addon-knobs/react';

import Button from './Button';

export const Primary = () => {
  const sizeKnob = select(
    'size',
    ['', 'sm', 'md', 'lg'],
  );
  const variantKnob = select(
    'variant',
    [
      '',
      'primary',
      'secondary',
      'tertiary',
      'link',
    ],
  );
  const textKnob = text('CTA Text', 'CTA Text');

  return (
    <Button size={sizeKnob} variant={variantKnob}>{textKnob}</Button>
  );
};

export default {
  title: 'Components/Button',
};
