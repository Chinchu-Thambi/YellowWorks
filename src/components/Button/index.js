import { Link } from 'rebass/styled-components';
import { Link as InnerLink } from 'gatsby';
import styled from 'styled-components';

import Button from './Button';

export default Button;

// TODO: Remove references to exports that use DeprecatedButton
const DeprecatedButton = styled(Link)`
  -webkit-appearance: inherit;
  display: inline-block;
  background: ${({ theme }) => theme.colors.primary};
  padding: 20px 40px;
  border-radius: 40px;
  text-decoration: none;
  text-transform: uppercase;
  color: ${({ theme }) => theme.colors.brand};
  font-size: ${({ theme }) => theme.fontSizes[1]};
  font-family: ${({ theme }) => theme.fonts.sans};
  cursor: pointer;
  text-align: center;
  border: 0;
`;

export const LinkButton = styled(DeprecatedButton)`
  && {
    background: transparent;
    display: flex;
    border: 3px;
    color: ${({ theme }) => theme.colors.cerulean};
  }
`;

export const InnerLinkButton = styled(InnerLink)`
  text-decoration: none;
  color: ${({ theme }) => theme.colors.cerulean};
  display: inline-block;
  padding: ${({ theme }) => theme.space[2]};
  font-weight: ${({ theme }) => theme.fontWeight[0]};
`;

export const SignInHeaderButton = styled(Button)`
  && { 
    border: 1px solid ${({ theme }) => theme.palette.contrast[3]};
    color: ${({ theme }) => theme.palette.contrast[2]};
    background: transparent;
    height: 40px ;
    margin: ${({ theme }) => theme.space[0]};
    padding: ${({ theme }) => theme.space[0]};
    padding-right: 16px;
    font-weight: ${({ theme }) => theme.fontWeight[1]};
    font-size: ${({ theme }) => theme.fontSizes[1]};
    white-space: nowrap;
}
  }

  &:hover {
    background: ${({ theme }) => theme.palette.contrast[5]};
  }

  svg {
    margin-right: ${({ theme }) => theme.space[2]};
    height: calc(100% + 2px);
    width: auto;
    position: relative;
    left: -1px;
  }
`;
