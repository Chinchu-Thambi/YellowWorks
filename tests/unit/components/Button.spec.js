/* globals describe, it, expect */

import React from 'react';
import { render, mount } from 'enzyme';

import Button from '../../../src/components/Button';

describe('buttons', () => {
  const theme = {
    colors: {
      primary: 'primaryColor',
      brand: 'brandColor',
      white: 'white',
    },
    palette: {
      base: [''],
      contrast: [],
    },
    fonts: {
      sans: 'fontSans',
    },
    fontSizes: [0, 1, 2, 3, 4, 5, 6],
  };

  it('exports a primary button', () => {
    const wrapper = render(<Button theme={theme} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('exports a secondary button', () => {
    const wrapper = render(<Button variant="secondary" theme={theme} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Button', () => {
    it('renders a <button> by default', () => {
      const wrapper = mount(<Button>Hello</Button>);
      expect(wrapper.text()).toBe('Hello');
      expect(wrapper.find('button')).toBeTruthy();

      wrapper.unmount();
    });

    it('renders a normal link <a> tag when passing "href" prop', () => {
      const wrapper = mount(<Button href="http://google.com">Link</Button>);
      expect(wrapper.text()).toBe('Link');
      expect(wrapper.find('a')).toBeTruthy();
      expect(wrapper.find('a').prop('href')).toBe('http://google.com');

      wrapper.unmount();
    });

    it('renders a Gatsby link when passing "to" prop', () => {
      const wrapper = mount(<Button to="/">Gatsby</Button>);
      expect(wrapper.text()).toBe('Gatsby');
      expect(wrapper.find('a')).toBeTruthy();
      expect(wrapper.find('a').prop('href')).toBe('/');

      wrapper.unmount();
    });

    it('renders with given extra props', () => {
      const wrapper = mount(
        <Button
          variant="primary"
          href="#"
          target="_blank"
          disabled
        >Link with props
        </Button>,
      );
      expect(wrapper.find('a')).toBeTruthy();
      expect(wrapper.find('a').prop('href')).toBe('#');
      expect(wrapper.find('a').prop('target')).toBe('_blank');

      wrapper.unmount();
    });
  });
});
