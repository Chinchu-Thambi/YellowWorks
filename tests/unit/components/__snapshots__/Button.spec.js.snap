// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buttons exports a primary button 1`] = `
<button
  class="sc-dlnjPT liGRPc"
/>
`;

exports[`buttons exports a secondary button 1`] = `
<button
  class="sc-dlnjPT imAVyc"
/>
`;
